examples/notebooks/get-started/learn_beam_windowing_by_doing.ipynb

```diff
@@ -0,0 +1,1836 @@
+        "\n",
+        "# 3. Put All Together\n",
+        "\n",
+        "Section 2 uses the toy data to go through three different windowing types. Now it is time to analyze the real data (`data/air_quality.csv`).\n",
+        "\n",
+        "Can you build one Beam pipeline to compute all AQI values for these windows?\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "plMjLuh-lKzr"
+      },
+      "outputs": [],
+      "source": [
+        "#@title Edit This Code Cell\n",
+        "..."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 15,
+      "metadata": {
+        "cellView": "form",
+        "id": "t6vBMax0bubN"
+      },
+      "outputs": [],
+      "source": [
+        "#@title Expand to see the answer\n",
+        "import apache_beam as beam\n",
+        "import csv\n",
+        "import time\n",
+        "import apache_beam as beam\n",
+        "from datetime import timedelta\n",
+        "import apache_beam.runners.interactive.interactive_beam as ib\n",
+        "\n",
+        "def to_unix_time(time_str: str, time_format='%Y-%m-%d %H:%M:%S') -> int:\n",
+        "  \"\"\"Converts a time string into Unix time.\"\"\"\n",
+        "  time_tuple = time.strptime(time_str, time_format)\n",
+        "  return int(time.mktime(time_tuple))\n",
+        "\n",
+        "def human_readable_window(window) -> str:\n",
+        "  \"\"\"Formats a window object into a human readable string.\"\"\"\n",
+        "  if isinstance(window, beam.window.GlobalWindow):\n",
+        "    return str(window)\n",
+        "  return f'{window.start.to_utc_datetime()} - {window.end.to_utc_datetime()}'\n",
+        "\n",
+        "@beam.ptransform_fn\n",
+        "def OutputWindowInfo(pcollection):\n",
+        "  \"\"\"Output the Window information with AQI in that window.\"\"\"\n",
+        "  class GetAQI(beam.DoFn):\n",
+        "    def process(self, mean_elements, window=beam.DoFn.WindowParam):\n",
+        "      yield human_readable_window(window), mean_elements\n",
+        "  return (\n",
+        "    pcollection\n",
+        "    | 'Count elements per window' >> beam.combiners.Mean.Globally().without_defaults()\n",
+        "    | 'Output counts info' >> beam.ParDo(GetAQI())\n",
+        "  )\n",
+        "\n",
+        "\n",
+        "def parse_file(element):\n",
+        "  file = csv.reader([element], quotechar='\"', delimiter=',',\n",
+        "                    quoting=csv.QUOTE_ALL, skipinitialspace=True)\n",
+        "  for line in file:\n",
+        "    return line\n",
+        "\n",
+        "p =  beam.Pipeline()\n",
+        "\n",
+        "# get the data\n",
+        "read_text = (p | 'Read input file' >> beam.io.ReadFromText(\"data/air_quality.csv\",\n",
+        "                                                  skip_header_lines=1)\n",
+        "| 'Parse file' >> beam.Map(parse_file)\n",
```

Comment:
nit: This formatting looks a bit strange in this section, since typically there is an indent in the text below opened brackets.

Also multiple long lines throghout the notebook.

Reply:
Formatting the cells is so painful. Colab does not provide the easy way to do that. I reformatted this cell and cleaned up some long lines (without a good formatting tool).